src/server/src/main/java/io/cassandrareaper/service/StreamManager.java

```diff
@@ -0,0 +1,204 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cassandrareaper.service;
+
+import io.cassandrareaper.AppContext;
+import io.cassandrareaper.ReaperException;
+import io.cassandrareaper.core.Cluster;
+import io.cassandrareaper.core.Node;
+import io.cassandrareaper.core.Stream;
+import io.cassandrareaper.core.StreamSession;
+import io.cassandrareaper.jmx.JmxProxy;
+import io.cassandrareaper.jmx.StreamStatusHandler;
+import io.cassandrareaper.storage.StreamCache;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import javax.management.openmbean.CompositeData;
+
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.Futures;
+import org.apache.cassandra.streaming.ProgressInfo;
+import org.apache.cassandra.streaming.SessionInfo;
+import org.apache.cassandra.streaming.management.ProgressInfoCompositeData;
+import org.apache.cassandra.streaming.management.SessionInfoCompositeData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class implements the functionality needed for receiving,
+ * handling and serving streaming-related notifications.
+ */
+public final class StreamManager implements StreamStatusHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamManager.class);
+  private final AppContext context;
+  private final ExecutorService executor = Executors.newFixedThreadPool(5);
+  private final StreamCache streamSessions = new StreamCache();
+
+  private StreamManager(AppContext context) {
+    this.context = context;
+  }
+
+  public static StreamManager create(AppContext context) {
+    return new StreamManager(context);
+  }
+
+  /**
+   * Connects a JmxProxy to each node in the given cluster
+   * and register notification handlers for streaming events.
+   */
+  public void addListeners(Cluster cluster) throws ReaperException {
+    final int jmxTimeout = context.config.getJmxConnectionTimeoutInSeconds();
+    try {
+      JmxProxy seedJmxProxy = context.jmxConnectionFactory.connectAny(cluster, jmxTimeout);
+
+      List<Callable<JmxProxy>> connectionTasks = seedJmxProxy.getLiveNodes()
+          .stream()
+          .map(hostName -> Node.builder().withCluster(cluster).withHostname(hostName).build())
+          .map(node -> connectJmxTask(node))
+          .collect(Collectors.toList());
+
+      List<JmxProxy> nodeProxies = Lists.newArrayList();
+      List<Future<JmxProxy>> futures = executor.invokeAll(connectionTasks);
+      for (Future<JmxProxy> future : futures) {
+        nodeProxies.add(future.get());
+      }
+
+      nodeProxies.forEach(nodeJmxProxy -> nodeJmxProxy.addStreamStatusHandler(this));
```

Comment:
How do you handle the case where a connection is lost to a specific node ? Will the handler be recreated when we create a new connection to this node ?

Reply:
I guess the best way to do this would be to subscribe in the JmxConnectionFactory instead.